lib/icinga/apiactions.cpp

```diff
@@ -555,5 +555,163 @@ Dictionary::Ptr ApiActions::GenerateTicket(const ConfigObject::Ptr&,
 Dictionary::Ptr ApiActions::ExecuteCommand(const ConfigObject::Ptr& object,
 	const Dictionary::Ptr& params)
 {
-	return ApiActions::CreateResult(501, "Not implemented");
+	Checkable::Ptr checkable = static_pointer_cast<Checkable>(object);
+	if (! checkable)
+		return ApiActions::CreateResult(404, "Can't schedule a command execution for non-existent object.");
+
+	ObjectLock oLock (checkable);
+
+	Host::Ptr host;
+	Service::Ptr service;
+	tie(host, service) = GetHostService(checkable);
+
+	String endpoint = "$command_endpoint$";
+	if (params->Contains("endpoint"))
+		endpoint = params->Get("endpoint");
+
+	MacroProcessor::ResolverList resolvers;
+	if (params->Contains("macros")) {
+		if (params->Get("macros").IsObjectType<Dictionary>())
+			resolvers.emplace_back("override",HttpUtility::GetLastParameter(params, "macros"));
```

Comment:
You're checking whether `params->Get("macros")` is a Dictionary, but using `HttpUtility::GetLastParameter(params, "macros")` then. Unify this to `HttpUtility::GetLastParameter(params, "macros")` and call that only once. (`params->Contains("macros")` is OK.)